### PR TITLE
Add docs for Channel

### PR DESCRIPTION
**Pull request by j8r**
Fixes https://github.com/crystal-lang/crystal/issues/5801

I'm not very familiar with channels and its concepts, this docs can be improved.

### CODE REVIEW

File: src/channel.cr [hunk 1]
```diff
@@ -1,5 +1,19 @@
 require "fiber"
 
+# Channels enables to execute multiple tasks concurrently.
```
Comment (r00ster91) [inline src/channel.cr]: No, channels are for communicating between fibers. They don't execute multiple tasks concurrently, that's what fibers do.

File: src/channel.cr [hunk 2]
```diff
@@ -1,5 +1,19 @@
 require "fiber"
 
+# Channels enable concurrent communication between fibers.
+# ```
```
Comment (Sija) [inline src/channel.cr]:
```suggestion

# ```
```

File: src/channel.cr [hunk 3]
```diff
@@ -1,5 +1,19 @@
...
+# channel = Channel(Nil).new
+
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
#
```

File: src/channel.cr [hunk 4]
```diff
@@ -41,10 +55,18 @@ abstract class Channel(T)
     @closed
   end
 
+  # Receive a value from the channel.
+  # ```
+  # channel = Channel(Int32).new
+  # channel.send(1)
+  # channel.receive # => 1
+  # ```
+  #
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
```

File: src/channel.cr [hunk 5]
```diff
@@ -1,5 +1,19 @@
...
+# Channels enable concurrent communication between fibers.
+# ```
+# channel = Channel(Nil).new
+
+# spawn do
+#   puts "Before send"
+#   channel.send(nil)
+#   puts "After send"
+# end
+
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
#
```

File: src/channel.cr [hunk 6]
```diff
@@ -41,10 +56,17 @@ abstract class Channel(T)
     @closed
   end
 
+  # Receive a value from the channel.
+  # ```
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
  #
  # ```
```

File: src/channel.cr [hunk 7]
```diff
@@ -1,5 +1,20 @@
 require "fiber"
 
+# Channels enable concurrent communication between fibers.
+
+# ```
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
#
# ```
```

File: src/channel.cr [hunk 8]
```diff
@@ -127,61 +151,71 @@ abstract class Channel(T)
     ReceiveAction.new(self)
   end
 
-  # :nodoc:
```
Comment (z64) [inline src/channel.cr]: Docs comments on these are fine, but these look like implementation details that should retain the `:nodoc:` flag?

File: src/channel.cr [hunk 9]
```diff
@@ -41,10 +57,18 @@ abstract class Channel(T)
...
+  # ```
+  # channel = Channel(Int32).new
+  # channel.send(1)
+  # channel.receive # => 1
+  # ```
   def receive
     receive_impl { raise ClosedError.new }
   end
 
+  # Receive a value from the channel, if any.
```
Comment (z64) [inline src/channel.cr]: The difference between `receive` and `receive?` is that `receive?` will not raise an exception if the channel is closed. Other `Channel` methods will raise exceptions if it is closed; I recommend reading [`channel_spec.cr`](https://github.com/crystal-lang/crystal/blob/master/spec/std/channel_spec.cr) and grep for `expect_raises` to better understand when this is expected to occur, and maybe mention this in some docs ("Raises `Channel::ClosedError` if the channel is closed", etc..)

File: src/channel.cr [hunk 10]
```diff
@@ -1,5 +1,21 @@
 require "fiber"
 
+# Channels enable concurrent communication between fibers.
+
```
Comment (Sija) [inline src/channel.cr]:
```suggestion
#
```

File: src/channel.cr [hunk 11]
```diff
@@ -1,5 +1,21 @@
 require "fiber"
 
+# Channels enable concurrent communication between fibers.
```
Comment (straight-shoota) [inline src/channel.cr]:
```suggestion
# A `Channel` enables concurrent communication between fibers.
```

Singular should be used to describe a singular entity. The plural could be misinterpreted that you'd need multiple channels for concurrent communication.

File: src/channel.cr [hunk 12]
```diff
@@ -1,5 +1,21 @@
...
+# ```
+# channel = Channel(Nil).new
+#
+# spawn do
+#   puts "Before send"
+#   channel.send(nil)
+#   puts "After send"
+# end
+#
+# puts "Before receive"
```
Comment (straight-shoota) [inline src/channel.cr]: The messages could be numbered to illustrate control flow.
```suggestion
# puts "1. Before receive"
```

File: src/channel.cr [hunk 13]
```diff
@@ -41,10 +57,18 @@ abstract class Channel(T)
     @closed
   end
 
+  # Receive a value from the channel.
```
Comment (straight-shoota) [inline src/channel.cr]:
```suggestion
  # Receives a value from the channel.
  # If there is a value waiting, it is returned immediately. Otherwise, this method blocks until a value is sent to the channel.
  #
  # Raises `ClosedError` if the channel is closed or closes while waiting for receive.
```

File: src/channel.cr (hunk 9 above)
Comment (straight-shoota) [reply src/channel.cr]: Yes, raised exceptions and other error behaviour (like returning `nil`) should be explicitly mentioned.

```suggestion
  # Receives a value from the channel.
  # If there is a value waiting, it is returned immediately. Otherwise, this method blocks until a value is sent to the channel.
  # 
  # Returns `nil` if the channel is closed or closes while waiting for receive.
```

Ditto for `#send`.

File: src/channel.cr (hunk 8 above)
Comment (straight-shoota) [reply src/channel.cr]: I don't think these structs are part of the public API. Nor should they. Adding comments is great, but please keep `:nodoc:`.

Comment (j8r) [reply src/channel.cr]: I remove it to add a comment, and I forget to put it again.

File: src/channel.cr [hunk 14]
```diff
@@ -1,5 +1,21 @@
 require "fiber"
 
+# A `Channel` enables concurrent communication between fibers.
+#
+#
```
Comment (Sija) [inline src/channel.cr]: These suggestions are really tiresome, desired outcome should be:

```cr
# A `Channel` ...
#
# ```
# ...
# ```
```

i.e. with 1 empty line between the doc comment and example below.

File: src/channel.cr (hunk 8 above)
Comment (ysbaddaden) [reply src/channel.cr]: Don't document the Action structs unless you're writing actual implementation docs that will help developers to understand the internal API design.

File: src/channel.cr (hunk 12 above)
Comment (ysbaddaden) [reply src/channel.cr]: I fail to see the point of the whole example. The `puts` don't mean anything and are unerelated to Channel. In fact a buffered and unbuffered channel will behave differently. I would prefer the example to spawn one fiber that will send a few messages then let the main fiber receive those messages in the order they were sent, which is the most important thing.

File: src/fiber.cr [hunk 15]
```diff
@@ -129,6 +129,7 @@ class Fiber
     @resume_event ||= Crystal::EventLoop.create_resume_event(self)
   end
 
+  # Set the current fiber to wait, in order to eventually execute others fibers.
```
Comment (ysbaddaden) [inline src/fiber.cr]: This is completely unrelated to documenting Channel. AFAIK developers are encouraged to use `sleep(0)` for the time being anyway.

**Comment by j8r** [comment]
@ysbaddaden I whish I could but I have not the experience to go into the details, and tell pros and cons of buffered vs unbuffered channels. Mine barely reach `Channel(T).new`, `spawn`, `send` and `receive` :sweat_smile: 
That's a reason why I opened this PR, I think it's important to do our best to document this major language feature.

**Comment by j8r** [comment]
@ysbaddaden I wish I could but I have not the experience to go into the details, and tell pros and cons of buffered vs unbuffered channels. Mine barely reach `Channel(T).new`, `spawn`, `send` and `receive` :sweat_smile: 
That's a reason why I opened this PR, I think it's important to do our best to document this major language feature.

**Comment by j8r** [comment]
@sdogruyol Maybe, I don't know when I would have enough experience/knowledge with Channels to further improve the documentation. I won't probably the best one for this task.
